syntax = "proto3";

package braggle;



message Attributes {
  map<string, string> misc = 1;
}

message Tag {
  string tagname = 1;
  Attributes attributes = 2;
  repeated Element children = 3;
}

message Element {
  oneof element_kind {
    string ref = 1;
    string text = 2;
    Tag tag = 3;
  }
}

message PartialServerState {
  int64 timestep = 1;
  string root_id = 2;
  map<string, Element> elements = 3;
}

message PollRequest {
  int64 since_timestep = 1;
}

message PollResponse {
  PartialServerState state = 1;
}


message ClickEvent {
  string element_id = 1;
}
message TextInputEvent {
  string element_id = 1;
  string value = 2;
}

message Interaction {
  oneof interaction_kind {
    ClickEvent click = 1;
    TextInputEvent text_input = 2;
  }
}

message InteractionRequest {
  Interaction interaction = 1;
}
message InteractionResponse {
}
